speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/DataCompressionMenu.cpp

```diff
@@ -0,0 +1,239 @@
+
+    // Write data compression type
+    char val[32];
+    sprintf(val, "%d", m_dataCompressionType);
+    GfParmSetStr(readParam, PRM_DATA_COMPRESSION, GFMNU_ATTR_SELECTED, val);
+
+    // Save custom compression to xml file
+    char buf[32];
+    sprintf(buf, "%d", m_customCompressionLevel);
+    GfParmSetStr(readParam, PRM_CUSTOM_COMPRESSION, GFMNU_ATTR_TEXT, buf);
+
+    // Write queued changes
+    GfParmWriteFile(nullptr, readParam, COMPRESSION_SCREEN_NAME);
+}
+
+/// @brief Saves the settings so the mediator (or future instances) can access them
+static void SaveSettings()
+{
+    int compressionLevel;
+    switch (m_customCompressionLevel)
+    {
+        case 0:  // none
+        {
+            compressionLevel = 1;
+            break;
+        }
+        case 1:  // Minimum
+        {
+            compressionLevel = 3;
+            break;
+        }
+        case 2:  // Medium
+        {
+            compressionLevel = 5;
+            break;
+        }
+        case 3:  // Maximum
+        {
+            compressionLevel = 9;
+            break;
+        }
+        default:  // Custom
+        {
+            compressionLevel = m_customCompressionLevel;
+            break;
+        }
+    }
+
+    SMediator* mediator = SMediator::GetInstance();
+    mediator->SetCompressionLevel(compressionLevel);
+
+    SaveSettingsToFile();
+}
+
+/// @brief Takes actions that need to be run on activation of the data selection screen
+static void OnActivate(void* /* dummy */)
+{
+    LoadSettings();
+}
+
+/// @brief Switches back to the data selection menu
+static void SwitchToDataSelectionMenu(void* /* dummy */)
+{
+    // go back to the researcher screen
+    GfuiScreenActivate(s_prevHandle);
+}
+
+/// @brief Saves the settings and then switched back to the data selection menu
+static void SaveAndGoBack(void* /* dummy */)
+{
+    SaveSettings();
+
+    SwitchToDataSelectionMenu(nullptr);
+}
+
+/// @brief        Sets the type of data compression
+/// @param p_info Information about the radio button list
+static void SelectCompression(tRadioButtonInfo* p_info)
+{
+    m_dataCompressionType = static_cast<DataCompressionType>(p_info->Selected);
+}
+
+/// @brief Handle input in the custom compression type textbox
+static void SetCompressionLevel(void*)
+{
+    char* val = GfuiEditboxGetString(s_scrHandle, m_customCompressionLevelControl);
+    char* endptr;
+    m_customCompressionLevel = (int)strtol(val, &endptr, 0);
+    if (*endptr != '\0')
+        std::cerr << "Could not convert " << val << " to long and leftover string is: " << endptr << std::endl;
+
+    // the compression level needs to be uneven
+    if (m_customCompressionLevel % 2 == 0)
+    {
+        m_customCompressionLevel++;
+    }
+
+    if (m_customCompressionLevel > MAX_COMPRESSION_LEVEL)
+    {
+        m_customCompressionLevel = MAX_COMPRESSION_LEVEL;
+    }
+    else if (m_customCompressionLevel < 1)
+    {
+        m_customCompressionLevel = 1;
+    }
```

Comment:
Use a Clamp function for this (or create one in RppUtils)

Reply:
ClampFloat exists and can probably be generalised to Clamp